Add JOSS paper draft to calibration module PR (#876)
CLIMADA-project/climada_python: doc/joss/calibration-module/paper.md

---
title: 'A Module for Calibrating Impact Functions in the Climate Risk Modelling Platform CLIMADA'
tags:
  - Python
  - climate risk
  - impact function
  - vulnerability
  - optimization
authors:
  - name: Lukas Riedel
    orcid: 0000-0002-4667-3652
    affiliation: "1, 2"
    corresponding: true
  - name: Chahan M. Kropf

>     orcid: 0000-0000-0000-0000

Author: @chahank Please add your ORCID

>     orcid: 0000-0000-0000-0000

Reviewer: ```suggestion
    orcid: 0000-0002-3761-2292
```

    affiliation: "1, 2"
  - name: Timo Schmid

>     orcid: 0000-0000-0000-0000

Author: @timschmi95 Please add your ORCID

Given probabilistic hazard event sets or weather forecasts, they enable the computation of associated risks or impacts, respectively.
Because impact functions are difficult to determine on larger spatial and temporal scales of interest, they are often calibrated using hazard, exposure, and impact data from past events.
We present a module for calibrating impact functions based on such data using established calibration techniques like Bayesian optimization.
It is implemented as submodule of the climate risk modeling platform CLIMADA and fully integrates into its workflow.

# Statement of need

Natural hazards like storms, floods, droughts, and extreme temperatures will be exacerbated by climate change.
In 2022 alone, natural disasters affected 185 million people and caused economic losses of more than US$ 200 billion [@cred_2022_2023].
Forecasting the impacts of imminent events as well as determining climate risk according to future socio-economic pathways is crucial for decision-making in societal, humanitarian, political, socio-economic, and ecological issues [@smith_human_2014].
One major source of uncertainty in such computations is the vulnerability [@rougier_risk_2013].

> Typically modeled as impact function that yields the percentage of affected exposure depending on hazard intensity, vulnerability is difficult to determine *a prioi*.

Author: I agree to the suggestion. We have the option to add footnotes. Wouldn't that be nicer to read? 🤔

> Typically modeled as impact function that yields the percentage of affected exposure depending on hazard intensity, vulnerability is difficult to determine *a prioi*.

Reviewer: Also good for me yes :).

Using hazard footprints, exposure, and recorded impacts from past events, researchers therefore employ calibration techniques to estimate unknown impact functions and use these functions for future risk projections or impact forecasts [@eberenz_regional_2021; @luthi_globally_2021; @welker_comparing_2021; @roosli_towards_2021; @kam_impact-based_2023; @Schmid2023a; @riedel_fluvial_2024].

> CLIMADA is a widely used framework for calculating weather- and climate-related impacts and risks [@aznar-siguan_climada_2019; @gabriela_aznar_siguan_2023_8383171].

Reviewer: aaa no need then. I thought the second citation was the second paper, but it is the code :D. As you prefer actually.

> CLIMADA is a widely used framework for calculating weather- and climate-related impacts and risks [@aznar-siguan_climada_2019; @gabriela_aznar_siguan_2023_8383171].

Author: I included your suggestion and the other paper 👍

> The aforementioned works calibrated impact functions with different optimization algorithms within the CLIMADA ecosystem, but lack a consistent implementation of these algorithms.

Author: "calibrated" is a verb here, so I think it is complete? 🤔

> The aforementioned works calibrated impact functions with different optimization algorithms within the CLIMADA ecosystem, but lack a consistent implementation of these algorithms.

Reviewer: You are grammatically right :D. But I find it hard to read. Maybe replace `works` with `studies` ?

> The aforementioned works calibrated impact functions with different optimization algorithms within the CLIMADA ecosystem, but lack a consistent implementation of these algorithms.

Author: Very good point indeed!

> The aforementioned works calibrated impact functions with different optimization algorithms within the CLIMADA ecosystem, but lack a consistent implementation of these algorithms.

Author: ```suggestion
The aforementioned studies calibrated impact functions with different optimization algorithms within the CLIMADA ecosystem, but lack a consistent implementation of these algorithms.
```

> OASIS, another well-adopted loss modeling framework, also does not feature tools for impact function calibration, and regards vulnerability solely as model input [@oasis].

Reviewer: There are quite a few other risk modeling platforms, but none that I know has a calibration module for impact functions so far. Shall I try to complete the list of other platforms here?

Author: Would be great!

Reviewer: I seems that the ones I knew have either stopped being open source or seem to have stopped development. Some others I know have just not really a clearly available open source package. So we can keep it to OASIS LMF.

Author: All the better! Thanks for checking

> OASIS, another well-adopted loss modeling framework, also does not feature tools for impact function calibration, and regards vulnerability solely as model input [@oasis].

Reviewer: ```suggestion
OASIS LMF, another well-adopted loss modeling framework, also does not feature tools for impact function calibration, and regards vulnerability solely as model input [@oasis].
```

Oasis is a bit a weird construct with several entities. Oasis is part commercial, part non-commercial, part open-source part not. So, here I think we should refer to the actual open source model which is the OASIS LMF.

With the proposed calibration module, we aim at unifying the approaches to impact function calibration, while providing an extensible structure that can be easily adjusted to particular applications.

> # Module Structure

Author: No, I would leave that out. We can argue for that in the other paper and then mention that we have an implementation ready, without further digging into it. Otherwise, it remains unclear why we would implement the "average" and "tragedy" optimization here. Also, I don't have a tutorial ready for them.

> # Module Structure

Reviewer: Yes, fair point.

Calibrating the parameters of impact functions can be a complex task involving highly non-linear objective functions, constrained and ambiguous parameters, and parameter spaces with multiple local optima.
To support different iterative optimization algorithms, we implemented an `Optimizer` base class from which concrete optimizers employing different algorithms are derived.
All optimizers receive data through a common `Input` object, which provides all necessary information for the calibration task.
This enables users to quickly switch between optimizers without disrupting their workflow.

> Pre-defined optimizers are based on the `scipy.optimize` module [@virtanen_scipy_2020] and the `BayesianOptimization` package [@nogueira_bayesian_2014].

Author: We already mention that the module is easily extensible with new optimizers. I would rather leave out your suggestion because it does align worse with the beginning of the next sentence ("The latter is...")

We provide an `BayesianOptimizerController` that iteratively explores and "exploits" the parameter space to find the global optimum and stop the optimization process then.
This only requires minimal user input for indicating the sampling density.

# Example Code

Given a hazard event set, and exposure, and associated impact data in the appropriate CLIMADA data structures, the calibration input can quickly be defined.
Users have to supply a cost function, the parameter space bounds, a function that creates an impact function from the estimated parameter set, and a function that transforms a CLIMADA ``Impact`` object into the same structure as the input impact data.

```python
import pandas as pd
from sklearn.metrics import mean_squared_log_error

from climada.hazard import Hazard
from climada.entity import Exposure, ImpactFuncSet, ImpfTropCyclone
from climada.engine import Impact
from climada.util.calibrate import (
    Input,
    BayesianOptimizer,
    BayesianOptimizerController,
    BayesianOptimizerOutputEvaluator,
    select_best,
)

def calibrate(hazard: Hazard, exposure: Exposure, impact_data: pd.DataFrame):
    """Calibrate an impact function with BayesianOptimizer"""

>     def impact_function_tropical_cyclone(v_half: float, scale: float) -> ImpactFuncSet:

Author: The code replicates the main part of the tutorial, and the plots are almost exactly the same. We can change it, but then it actually becomes harder to replicate the example. However, I could also change the tutorial to use `from_sigmoid_impf`.

>     def impact_function_tropical_cyclone(v_half: float, scale: float) -> ImpactFuncSet:

Reviewer: I would do that. In part because the impact function for TC is quite particular, and in part because of the near-future plans to move TC to the petals repo.

>     def impact_function_tropical_cyclone(v_half: float, scale: float) -> ImpactFuncSet:

Author: I can then actually showcase a calibration with `v_thresh`, `v_half` and constraints!

>     def impact_function_tropical_cyclone(v_half: float, scale: float) -> ImpactFuncSet:

Author: @chahank Actually, the sigmoid implementation in `from_sigmoid_impf` is really not useful. It has no threshold parameter, meaning that you can easily create a function where `mdr(0) != 0`. It would be more useful to generalize the Emanuel-style sigmoid definition, as we did in the other paper. But this will also require the definition of a new function. So, for the sake of simplicity, I would keep everything as is.

        """Return an impact function set for tropical cyclones given two parameters

        This assumes that the input data relates to tropical cyclone asset damages.
        """
        return ImpactFuncSet(
            [ImpfTropCyclone.from_emanuel_usa(v_half=v_half, scale=scale)]
        )

>     def aggregate_impact_country_level(impact: Impact) -> pd.DataFrame:
>         """Aggregate modeled impacts and return as DataFrame"""
>         return impact.impact_at_reg()

Author: Maybe just change the wording?
```suggestion
    def aggregate_impact(impact: Impact) -> pd.DataFrame:
        """Aggregate modeled impacts per region ID and return as DataFrame"""
        return impact.impact_at_reg()
```

>         return impact.impact_at_reg()

Reviewer: Perfect!
